articles/WideStrings.md

```diff
@@ -0,0 +1,264 @@
+
+Cons:
+
+* If the user-level code uses a different encoding, the user is responsible
+for doing a conversion to the ROS 2.0 defined encoding.  This can involve
+some runtime cost.
+
+**User defines the encoding**
+
+Pros:
+
+* No conversions necessary for user-to-ROS 2 API.
+
+Cons:
+
+* The user must always tell ROS 2.0 what encoding the data is in.
+* The encoding needs to be transmitted on the wire to the other side.
+* It is not clear how ROS 2.0 will transmit the encoding.  Fixed list?  Which
+encodings go into the list?
+
+While a user defined encoding looks somewhat attractive at first, the downsides
+to it make it difficult to determine exactly which encodings are in use.  The
+downsides also mean that any program that wants to parse ROS 2 messages (or bag
+files) may potentially have to deal with many different encodings.  Additionally,
+Unicode (and in particular, UTF-8) are very commonplace now, while other encodings
+are becoming more esoteric.  Even earlier Unicode (UTF-16 and UTF-32) encodings are
+essentially deprecated in favor of UTF-8.  For these reasons, the current design of
+the ROS 2.0 wide string is to use a UTF-8 encoding for all strings.
+
+### What does the API look like to a user of ROS 2.0?
+
+#### Python 3
+
+Python 3 splits byte arrays from strings pretty cleanly.  Thus, in Python 3, a
+string is a sequence of characters, where each character can take 1 or more bytes.
+Byte arrays are sequences of bytes.  Thus, the ROS 2.0 API for dealing with wide strings
+(such as the publish API) will take just a string type in, and encode it as utf-8.
+The ROS 2.0 API for dealing with regular strings will take a string type in, and
+encode it as ASCII.
+
+**Example**
+
+```
+import sys
+from time import sleep
+
+import rclpy
+
+from rclpy.qos import qos_profile_default
+
+from std_msgs.msg import WString
+
+def main(args=None):
+    if args is None:
+        args = sys.argv
+
+    rclpy.init(args)
+
+    node = rclpy.create_node('talker')
+
+    chatter_pub = node.create_publisher(WString, 'chatter', qos_profile_default)
+
+    msg = String()
+
+    i = 1
+    while True:
+        msg.data = 'Hello World: {0}'.format(i)
+        i += 1
+        print('Publishing: "{0}"'.format(msg.data))
+        chatter_pub.publish(msg)
+        # TODO(wjwwood): need to spin_some or spin_once with timeout
+        sleep(1)
+
+
+if __name__ == '__main__':
+    main()
+```
+
+Notice that this code looks almost identical to the same demo for String.  That's because
+in python, the characters in the string and the byte representation are cleanly
+separated.  Thus, the API looks exactly the same as it would for the String type.
+
+#### C++
+
+C++ doesn't have good built-in support for wide strings.  Using wchar_t is generally
+not a good option, as wchar_t has different sizes on different platforms (2 bytes on
+Windows, 4 bytes on Linux).  Thus, ROS 2.0 will define a new type to deal with wide strings.
```

Comment:
I actually have come to the idea that adding this new type for C++ is the whole point of the WideString exercise.  The fact is that today, you can deal with UTF-8 data in ROS1 *or* ROS2 in 2 different ways: either by passing it through a byte array, or by passing it through the String type.  Neither place constraints on what bytes can be placed in them, and both of them will carry the data over to the destination, where it can be pulled back out.

Where WStrings come into play is putting some structure around this and guaranteeing that all parties agree on the encoding and agree on an API for dealing with them.  Since the story around UTF-8 in C++ is generally so weak, this would be an opportunity for ROS2 to make things easier for users, both at the API level, and when dealing with the data (as they'll always know what the encoding in).

So with all of that being said, we don't really *need* a WString type; it just makes dealing with certain things easier.  Given that, I'm not quite sure where we go with this.